Add "Direction" column to peers tab

jonatack opened the pull request:
![Screenshot from 2021-04-24 13-15-54](https://user-images.githubusercontent.com/2415484/115995575-4df86280-a5cb-11eb-84a6-7f2d3143b980.png)

This was initially proposed in #179 and saw an enthusiastic reception with several ACKs. One contributor was a weak NACK due to the horizontal space and so this was temporarily dropped, but since then #202 (and soon #256 and #293) resolve the issue. Moreover, the current state of displaying only a connection `Type` column without a connection `Direction` column is confusing (e.g. https://github.com/bitcoin/bitcoin/issues/21747#issuecomment-825072559), particularly as both the direction and the type are displayed in the peer details area with the `Direction/Type` row.

This patch fixes it and completes the initial proposal in #179 by adding a `Direction` column, making the peers tab the same as the `Direction/Type` row in the peer details and the direction and type columns in our other user-facing peer connections table in `-netinfo`.

Users can now sort the peers table by direction. The default sort is set to inbound, then outbound.

This patch also makes the direction and type translations the same for the 3 places where they are used (the peers column, the peer details row, and its tooltip), adds translator comments to link them together, and adds QStringBuilder fast QString concatenation with the `%` operator.

jarolrod reviewed (commented):
Concept ACK

I think this is an improvement over the current little arrows next to the address. But, this improvement is dependent on getting translations done. 

What if we kept the arrow next to the `Inbound` and `Outbound` text. That way, if there is no translation available, the arrows provide some context to go on.

hebasto commented on src/qt/peertablemodel.cpp:
```diff
@@ -161,8 +163,9 @@ QVariant PeerTableModel::data(const QModelIndex &index, int role) const
-            return QString(rec->nodeStats.fInbound ? "↓ " : "↑ ") + QString::fromStdString(rec->nodeStats.addrName);
+            return QString::fromStdString(rec->nodeStats.addrName);
+        case Direction:
+            return QString(rec->nodeStats.fInbound ? "Inbound" : "Outbound");
```
Also we could add translator comments now (see #287 for details).

jonatack replied:
Done.

jonatack commented:
> What if we kept the arrow next to the `Inbound` and `Outbound` text. That way, if there is no translation available, the arrows provide some context to go on.

Thanks for reviewing! Pushed with this change. WDYT?

before

![Screenshot from 2021-04-24 13-15-54](https://user-images.githubusercontent.com/2415484/115957598-f5559680-a4f2-11eb-9bdc-8c4965e79751.png)

after

![Screenshot from 2021-04-24 12-44-27](https://user-images.githubusercontent.com/2415484/115957619-0a322a00-a4f3-11eb-8c8a-d03786d7a520.png)

jonatack commented:
> What if we kept the arrow next to the `Inbound` and `Outbound` text. That way, if there is no translation available, the arrows provide some context to go on.

Thanks for reviewing! Pushed with this change. WDYT? If it was just me I would probably leave out the arrow as it's cleaner to my eye, but I don't have a strong opinion. 

before

![Screenshot from 2021-04-24 13-15-51](https://user-images.githubusercontent.com/2415484/115957821-271b2d00-a4f4-11eb-9b33-9e1aeb76c5cd.png)

after

![Screenshot from 2021-04-24 13-51-46](https://user-images.githubusercontent.com/2415484/115957732-a5c39a80-a4f3-11eb-8d66-2da89940c0a3.png)

RandyMcMillan commented:
@jonatack - In this previous PR #135 I did something similar...

![](https://user-images.githubusercontent.com/152159/100163372-87de0500-2e83-11eb-8d84-27afab6ffd07.png)

The goal was to actually embed unicode arrows in the column and not use text for direction at all.

This may be better visually and save column space.

The column sorting function should sort by unicode value with little to no additional work.

I just wanted to mention that since this PR is focusing on the direction column.

jonatack commented:
Thanks -- yes, I know but don't prefer a column with just arrows.

More objectively, there is an issue described in the PR description above: the Inbound/Outbound direction is displayed in the peer details area (the `Direction/Type` row) but not in the peers windows itself, which is confusing to users. Now that space is less of an issue thanks to #202 and #256, we can complete what was started in #179 and properly display the Direction and the Type next to each other as we are already doing in the peer details.

RandyMcMillan commented:
#293 should fit with #202 #256 & #179 as well...

What do you think leave the Ampersands or take them out?

![](https://user-images.githubusercontent.com/152159/115975495-42f1f380-a533-11eb-8a47-6d29a84bf28b.png)

jonatack commented:
> What do you think leave the Ampersands or take them out?

The current space-ampersand-space, or comma-space, seem like good separators.

jonatack commented:
Update: after adding the arrows yesterday, I removed them today because the `Direction` and `Type` columns should display the same thing as the `Direction/Type` details row. In a new second commit, I made the direction and type translations the same for the three places where they are used (peers column, peer details row, and its tooltip) and added translator comments to link them together. It's not clear to me, however, if the translator comments should refer to places in the codebase or to where the translations are displayed to users.

Talkless commented on src/qt/rpcconsole.cpp:
```diff
@@ -462,15 +462,17 @@ RPCConsole::RPCConsole(interfaces::Node& node, const PlatformStyle *_platformSty
-        tr("Outbound Manual: added using RPC %1 or %2/%3 configuration options")
+        //: The connection direction (Inbound/Outbound) is also used in ConnectionTypeToQString() and PeerTableModel::data().
+        //: The connection types (Full Relay, Block Relay, Manual, Feeler, Address Fetch) are also used in ConnectionTypeToQString().
+        tr("Inbound") + ": " + tr("initiated by peer"),
```
nit: Now there's lot's of string concatenations, maybe worth using QStringBuilder's `%` and wrapping " " and ": " with `QLatin1String` ?

jonatack replied:
Thanks @Talkless. Added QLatin1String constants in 73cf0fdc8edacf9294f9facc8248fb078298b2d7. Let me know if you think using `%` would still improve the diff and what you'd suggest.

Talkless reviewed (approved):
tACK d04ebdaab74493bb0cba79fbb81b8225e8bbb8de, tested on Debian Sid with Qt 5.15.2.

Talkless commented on src/qt/addressbookpage.cpp:
```diff
@@ -16,6 +16,7 @@
 #include <qt/platformstyle.h>
 
 #include <QIcon>
+#include <QLatin1String>
```
Uhm, I'm a bit lost, why do we have QLatin1String is included a lot in this PR?

jonatack replied:
Added the header where it is used but header is missing.

Talkless commented on src/qt/rpcconsole.cpp:
```diff
@@ -52,6 +54,9 @@ const int INITIAL_TRAFFIC_GRAPH_MINS = 30;
 const QSize FONT_RANGE(4, 40);
 const char fontSizeSettingsKey[] = "consoleFontSize";
 
+const QLatin1String COLON_SPACE{": "};
```
Interesting idea. It actually could be `constexpr`.

jonatack replied:
That was the first thing I tried but the compiler doesn't like it.

Talkless replied:
That's strange, QLatin1String's constructors are constexpr-marked since at least 5.3: https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qstring.h?h=5.3#n89

Sure that wasn't type or something? What kind of message was that? And what compiler?

jonatack replied:
I'm using Qt version 5.15.2 and according to doc/dependencies.md Bitcoin Core uses 5.12.10 and supports back to 5.9.5.

jonatack replied:
This seemed relevant: https://stackoverflow.com/questions/56201976/qt-vs-constexpr-string-literal

Clang 9 and GCC 10.2.1:
```
  CXX      qt/libbitcoinqt_a-rpcconsole.o
qt/rpcconsole.cpp:58:32: error: constexpr variable 'COLON_SPACE' must be initialized by a constant expression
static constexpr QLatin1String COLON_SPACE{": "};
                               ^~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:91:93: note: non-constexpr function 'strlen' cannot be used in a constant expression
    Q_DECL_CONSTEXPR inline explicit QLatin1String(const char *s) noexcept : m_size(s ? int(strlen(s)) : 0), m_data(s) {}
                                                                                            ^
qt/rpcconsole.cpp:58:32: note: in call to 'QLatin1String(&": "[0])'
static constexpr QLatin1String COLON_SPACE{": "};
                               ^
qt/rpcconsole.cpp:59:32: error: constexpr variable 'SPACE' must be initialized by a constant expression
static constexpr QLatin1String SPACE{" "};
                               ^~~~~~~~~~
/usr/include/x86_64-linux-gnu/qt5/QtCore/qstring.h:91:93: note: non-constexpr function 'strlen' cannot be used in a constant expression
    Q_DECL_CONSTEXPR inline explicit QLatin1String(const char *s) noexcept : m_size(s ? int(strlen(s)) : 0), m_data(s) {}
                                                                                            ^
qt/rpcconsole.cpp:59:32: note: in call to 'QLatin1String(&" "[0])'
static constexpr QLatin1String SPACE{" "};
                               ^
2 errors generated.
```

Talkless commented on src/qt/rpcconsole.cpp:
```diff
@@ -36,12 +36,14 @@
 #include <QScrollBar>
 #include <QSettings>
 #include <QString>
+#include <QStringBuilder>
```
Not sure if it helps only by including (without using `operator %`), unless whole bitcoin-qt would be built with `QT_USE_QSTRINGBUILDER`.

jonatack replied:
Added because this file already uses `%`.

Talkless replied:
I see, QLatin1String and QStringBuilder included because where simply _missing_, not because of new usage. New usage would be nice though, but I don't want to be too-QString-nazi..

jonatack replied:
Thanks for the nudge, @Talkless. Looked up https://wiki.qt.io/Using_QString_Effectively and updated. Please let me know what you think.

jonatack commented:
Rebased due to the merge of #18. Thanks for the reviews @Talkless!

prayank23 commented:
Concept ACK

prayank23 commented on src/qt/peertablemodel.cpp:
```diff
@@ -113,8 +113,10 @@ QVariant PeerTableModel::data(const QModelIndex &index, int role) const
+            return QString::fromStdString(rec->nodeStats.addrName);
+        case Direction:
+            //: Connection direction. Also used in ConnectionTypeToQString() and CONNECTION_TYPE_DOC.
+            return QString(rec->nodeStats.fInbound ? tr("Inbound") : tr("Outbound"));
```
Can we use `IN` and `OUT` in caps instead of Inbound and Outbound?

jarolrod replied:
Here is how that would look, not a bad idea. But, a) I think this change is ok as is b) would like to hear other opinions.

![Screen Shot 2021-04-30 at 3 09 34 AM](https://user-images.githubusercontent.com/23396902/116660814-89938380-a961-11eb-9bd4-a37ee76ab85e.png)

jonatack replied:
The goal is to unify the peers tab Direction and Type columns with the Direction/Type row in the peer details sidebar, so if the side bar displays "Outbound Full Relay", the main table also displays "Outbound Full Relay" with the ability to sort by either Direction or Type.

jonatack replied:
I tried to communicate it in the PR description.

jonatack replied:
(I also tried "In/Out" while working on this PR, and the horizontal space isn't reduced unless the column header is "Dir" instead of "Direction". But the other headers are full words.)

prayank23 commented:
> ![Screenshot from 2021-04-24 13-15-54](https://user-images.githubusercontent.com/2415484/115995575-4df86280-a5cb-11eb-84a6-7f2d3143b980.png)

How do you have 3 Block Relay connections in this screenshot? Isn't [MAX 2](https://github.com/bitcoin/bitcoin/blob/66fd3b28e85c167f3955b5603496daf8d91abcad/src/net.h#L68)?

jarolrod reviewed (commented):
ACK 2bf3f5aee014ab4e0ff72e6bdbb33ada02dec314

This patch works well and is a UX improvement. Tested on macOS 11.3 Qt 5.15.2

Thanks for working on this and adding in translator comments.

![Screen Shot 2021-04-30 at 2 59 31 AM](https://user-images.githubusercontent.com/23396902/116660066-7df38d00-a960-11eb-81fa-8252e38e4765.png)

jonatack commented:
> How do you have 3 Block Relay connections in this screenshot? Isn't [MAX 2](https://github.com/bitcoin/bitcoin/blob/66fd3b28e85c167f3955b5603496daf8d91abcad/src/net.h#L68)?

We very often have a short-lived third extra block relay connection since the merge of https://github.com/bitcoin/bitcoin/pull/19858. 

See this open PR that updates our docs about this: https://github.com/bitcoin/bitcoin/pull/21709/files.

hebasto commented:
Concept ACK on removing direction info from the "Address" column.

@luke-jr wrt your comment https://github.com/bitcoin-core/gui/pull/179#issuecomment-762417066 and new [circumstances](https://github.com/bitcoin-core/gui/pull/289#issue-622087016), did you change your mind?

jonatack commented:
Up for grabs.

prayank23 commented:
I can work on this. But I am not sure why is this closed and "Up for grabs" because @jonatack already made changes required and reviews looked positive as well.

hebasto commented:
@prayank23 
> I can work on this.

You're also welcome to join the `##bitcoin-core-gui` (with two #) channel on IRC :)

jonatack commented:
Hi @prayank23, sorry for the confusion.  I opened this to finish what was started with #179 but decided to not spend further time on GUI/design/QT/UI/translation things.  Everything takes time to do well, and discussing each word, arrow, indentation, translation, and consistency question and explanation until everyone is happy is completely normal, but it takes bandwidth in an area that isn't my focus.  Cheers.

jarolrod commented:
@hebasto this is no longer up for grabs as it was picked up in #317
